File: cpp/ampls/examples/multipleModels.cpp
#include <string>
#include <exception>
#include <cassert>
#include "ampls/ampls.h"
#include "test-config.h" // for MODELS_DIR

#include "ampl/ampl.h"


/// This example shows how to create AMPL models using the AMPL API and detect various 
/// errors that can arise when importing them to AMPLS

void makeAmplModel(ampl::AMPL &ampl , int numvars, bool unfeasible, bool unbounded=false) {
  assert(!(unfeasible && unbounded));
  ampl.eval("set varIndexes within Integers; set constraintIndexes within Integers;"
    "param varCosts{ varIndexes };"
    "param rightHandSides{ constraintIndexes };"
    "var x{ i in varIndexes } >= 0;"
    "maximize objectiveFunction : sum{ i in varIndexes } varCosts[i] * x[i];");
    

  ampl::DataFrame d(1, { "varIndexes", "varCosts" });
  for (int i = 0; i < numvars; i++)
    d.addRow(i, i);
  ampl.setData(d, "varIndexes");
  if (!unbounded)
  {
    ampl.eval(" subject to mainConstraints{ j in constraintIndexes } : x[j] + x[j + 1] <= rightHandSides[j];");
    ampl::DataFrame d2(1, { "constraintIndexes", "rightHandSides" });
    for (int i = 0; i < numvars - 1; i++)
      d2.addRow(i, unfeasible ? -1 : 1);
    ampl.setData(d2, "constraintIndexes");
  }
}

/// <summary>
/// Export the AMPL model to a (templated) AMPLS solver and solve it via AMPLS
/// Reimports the model in the AMPL API object and also returns the AMPLS object for
/// further analysis 
/// </summary>
template <class T> T solveModel(ampl::AMPL& ampl, const std::map<std::string, int>& options = {}) {
    auto model = ampls::AMPLAPIInterface::exportModel<T>(ampl);
    for (auto o : options)
    {
      try {
        model.setOption(o.first.c_str(), o.second);
      }
      catch (...) {}
    }
    model.optimize();
    ampls::AMPLAPIInterface::importModel(ampl, model);
    return model;
}
/// <summary>
/// Template function for solver-specific methods
/// </summary>
template <class T> void checkNativeStatus(T& model) {}
#ifdef USE_gurobi
void checkNativeStatus(ampls::GurobiModel& grbmodel) {
  // Use AMPLS wrapper to Gurobi native function
  int status = grbmodel.getIntAttr(GRB_INT_ATTR_STATUS);
  assert(status == GRB_OPTIMAL);

  // Use Gurobi native functions
  GRBgetintattr(grbmodel.getGRBmodel(), GRB_INT_ATTR_STATUS, &status);
  assert(status == GRB_OPTIMAL);
}
#endif
#ifdef USE_cplex
void checkNativeStatus(ampls::CPLEXModel& model) {
  // Use CPLEX native functions
  int status = CPXgetstat(model.getCPXENV(), model.getCPXLP());
  assert(status == CPX_STAT_OPTIMAL);
}
#endif
#ifdef USE_xpress
void checkNativeStatus(ampls::XPRESSModel& model) {
  // Use AMPLS wrapper to XPRESS native function
  int status = model.getIntAttr(XPRS_LPSTATUS);
  assert(status = XPRS_LP_OPTIMAL);
  
  // Use XPRESS native functions
  XPRSgetintattrib(model.getXPRSprob(), XPRS_LPSTATUS, &status);
  assert(status = XPRS_LP_OPTIMAL);
}
#endif


template <class T> void createAndSolveSimpleModel() {
  int numVars = 10;
  ampl::AMPL ampl;
  makeAmplModel(ampl, numVars, false);
  auto model = solveModel<T>(ampl);

  // Get status through AMPLS
  assert(model.getStatus() == ampls::Status::OPTIMAL);

  // Demonstrate solver native methods
  checkNativeStatus(model);

  // Num Vars is even->last var idx is odd->expect odd idx vars to be 1
  // Num Vars is odd->last var idx is even->expect even idx vars to be 1
  std::vector<double> expectedSolution(numVars);
  double expectedObjective = 0;
  for (int i = 0; i < numVars; i++) {
    expectedSolution[i] = std::abs(i % 2 - numVars % 2);
    expectedObjective += i * expectedSolution[i];
  }
  double solverObjective = model.getObj();
  assert(solverObjective == expectedObjective);
  printf("Completed Simple Model Test.\n\n");
}

template <class T> void createAndSolveTrivialModel() {
  ampl::AMPL ampl;
  ampl.eval("var x; maximize z:x; c:x=5;");
  auto model = solveModel<T>(ampl);
}

template <class T> void createAndSolveInfeasibleModel(bool presolve) {
    int numVars = 10;
    ampl::AMPL ampl;
    makeAmplModel(ampl, numVars, true);
    ampl.setIntOption("presolve", presolve ? 10 : 0); // 10 is the default number of presolve passes
    std::map<std::string, int> options = { {"outlev", 1}, {"iisfind", 1} };
    auto model = solveModel<T>(ampl, options);
    assert(model.getStatus() == ampls::Status::INFEASIBLE);
    printf("Completed Infeasible Model Test\n\n");
}

template <class T> void createAndSolveUnboundedModel(bool presolve) {
  int numVars = 10;
  ampl::AMPL ampl;
  makeAmplModel(ampl, numVars, false, true);
  ampl.setIntOption("presolve", presolve ? 100 : 0); // 10 is the default number of presolve passes
  std::map<std::string, int> options = { {"outlev", 1}};
  auto model = solveModel<T>(ampl, options);
  assert(model.getStatus() == ampls::Status::UNBOUNDED);
  printf("Completed Unbounded Model Test\n\n");
}

template <class T> void example() {

  createAndSolveInfeasibleModel<T>(false);

  // Normal model solver to completion
  createAndSolveSimpleModel<T>();
  bool caught = false;
  try {
    // This will fail because AMPL's presolve will block the 
    // export of the model
    createAndSolveInfeasibleModel<T>(true);
  }
  catch (const ampl::InfeasibilityException& e) {
    printf("Completed Infeasible presolved Model Test\n\n");
    printf("Caught exception:\n%s\n", e.what());
    caught = true;
  }
  assert(caught);
  caught = false;
  try {
    createAndSolveTrivialModel<T>();
  }
  catch (const ampl::PresolveException& e) {
    printf("Completed Trivial Model Test\n");
    printf("Caught exception:\n%s\n", e.what());
    caught = true;
  }
  assert(caught);

  

  // Unbounded models are exported either way
  createAndSolveUnboundedModel<T>(true);
  createAndSolveUnboundedModel<T>(false);


}

int main(int argc, char** argv) {
#ifdef USE_highs
  example<ampls::HighsModel>();
#endif
#ifdef USE_cplex
  example<ampls::CPLEXModel>();
#endif
#ifdef USE_xpress
  example<ampls::XPRESSModel>();
#endif
#ifdef USE_gurobi
  example<ampls::GurobiModel>();
#endif
#ifdef USE_cbcmp
  example<ampls::CbcModel>();
#endif

#ifdef USE_scip
  example<ampls::SCIPModel>();
#endif
#ifdef USE_copt
  //example<ampls::CoptModel>();
#endif
  return 0;
}

File: python/examples/multiple_models.py
import amplpy
import pandas as pd
from typing import Any, Dict

import amplpy_cplex as ampls
SOLVER = "cplex"

# Example description
# Shows how to create multiple models with AMPL, export an AMPLs instance
# for a specific solver, set some solver options and get status/solution, and get the results
# back in AMPL. It also shows how to use the native C API of various solvers


def makeAmplModel(num_vars :int=10, flip_objective:bool = False, infeasible:bool = False, 
                  presolve_level : bool = 10):
    """ Create an instance of AMPL and a model"""
    ampl = amplpy.AMPL()
    # Note that if a model is detected as infeasible by the AMPL presolver,
    # it will not be exported to the solver, and an exception will be thrown.
    ampl.set_option("presolve", presolve_level)
    
    ampl.eval("""
    set varIndexes within Integers;
    set constraintIndexes within Integers;

    param varCosts { varIndexes };
    param rightHandSides { constraintIndexes };

    var x {i in varIndexes} >=0;

    maximize objectiveFunction:
        sum { i in varIndexes } varCosts[i] * x[i];

    subject to mainConstraints { j in constraintIndexes } :
        x[j] + x[j+1] <= rightHandSides[j];"""
    )

    varIndexDf = pd.DataFrame(
        data = [(i,i if not flip_objective else -i) for i in range(num_vars)],
        columns = ["varIndexes","varCosts"]
    ).set_index("varIndexes")
    ampl.setData(varIndexDf,set_name = "varIndexes")
    
    constraintIndexesDf = pd.DataFrame(
        data = [(i,1 if not infeasible else -1) for i in range(num_vars-1)],
        columns = ["constraintIndexes","rightHandSides"]
    ).set_index("constraintIndexes")
    ampl.setData(constraintIndexesDf,set_name = "constraintIndexes")
    return ampl

def solveModel(ampl:amplpy.AMPL, options:Dict[str,Any] = {}, useNativeCall: bool=True):
    """Solves a model by exporting it to ampls and using either the generic optimize()
       function or the solver native C API"""
    model= ampl.to_ampls(SOLVER)
    for o, v in options.items():
        model.set_option(o,v) # Set options from
       # ampl driver (see for example https://dev.ampl.com/solvers/gurobi/options.html,  
       #  https://dev.ampl.com/solvers/xpress/options.html# )
       # Most of them will work the same way across solvers
    
    # For specific solvers there are also wrappers for the 
    # gurobi parameter accessors. See https://www.gurobi.com/documentation/10.0/refman/c_parameter_examples.html
    if SOLVER == "gurobi":
        model.set_param("Threads", 4) 

    # A third option is to use the solver C API directy, by getting access to the underlying 
    # gurobi model/environment pointer. This is suitabe to use functionalities provided
    # by the specific solver C API
    # The two statmeents below are therefore equivalent, the first uses the Gurobi C API directly,
    # the latter uses the (generic) function provided by ampls
    if useNativeCall:
        if SOLVER == "gurobi":
            grbnative = model.get_grb_model()
            ampls.GRBoptimize(grbnative)
        if SOLVER == "cplex":
            cplexnativeenv = model.get_cplex_env()
            cplexnativemodel = model.get_cplex_lp()
            ampls.CPXlpopt(cplexnativeenv, cplexnativemodel)
        if SOLVER == "xpress":
            xpressnative = model.getXPRSprob()
            ampls.XPRSlpoptimize(xpressnative, None)
    else:
        model.optimize()
    ampl.import_solution(model)
    return model



def createAndSolveSimpleModel(useNativeCall=True):
    num_vars = 10
    ampl = makeAmplModel(num_vars=num_vars)
    model = solveModel(ampl, useNativeCall=useNativeCall)
    
    print(f"Status is {model.get_status().name}")
    assert model.get_status() == ampls.Status.OPTIMAL

    # Num Vars is even -> last var idx is odd -> expect odd idx vars to be 1
    # Num Vars is odd -> last var idx is even -> expect even idx vars to be 1
    expectedSolution = [abs(i%2-num_vars %2) for i in range(num_vars )]    
    expectedObjective = sum( [i*expectedSolution[i] for i in range(num_vars )])    
    solverObjective = model.getObj()
    assert solverObjective == expectedObjective
    # Check that AMPL has received the correct objective
    assert solverObjective == ampl.getCurrentObjective().value()
    print("Completed Simple Model Test.")

def createAndSolveInfeasibleModel(presolve_level : bool = 10, useNativeCall=True):
    ampl = makeAmplModel(infeasible=True, presolve_level=presolve_level)
    # Set some solver options
    # Setting iisfind to 1 to find the Irreducible Infeasibility Subset
    model = solveModel(ampl, {"outlev" : 1, "iisfind" : 1}, useNativeCall=useNativeCall)
    print(f"Status is {model.get_status().name}")
    assert model.get_status() == ampls.Status.INFEASIBLE
    
    # Display IIS for infeasible model by converting it to a pandas dataframe
    print(ampl.get_data("_varname, _var.iis").to_pandas().set_index('_varname'))
    print(ampl.get_data("_conname, _con.iis").to_pandas().set_index('_conname'))
    print("Completed Infeasible Model Test.")


if __name__ == "__main__":
    try:
        # With presolve set to a high level, AMPL detects the infeasibilitty
        # and does not export the model. Throws an explainatory runtime error
        createAndSolveInfeasibleModel(10)
    except ampls.AMPLSolverException as e:
        print(e)
    # Turning off presolve makes ampl actually export the model
    # In the fucntion we'll set some options to find the source of the infeasibility
    createAndSolveInfeasibleModel(0)

    # Repeated solves below for feasible model:
    createAndSolveSimpleModel()
    createAndSolveSimpleModel()
    